### PR TITLE
fix: add support for localhost urls in markdown

### PR DESCRIPTION
**Pull request by allroundexperts**
This PR adds support for auto linking of local urls like `http://my-localhost`, `http://localhost:3000`

### Fixed Issues
$ https://github.com/Expensify/App/issues/17828

# Tests
1. What unit/integration tests cover your change? What autoQA tests cover your change?
`URL-test`, `ExpensiMark-HTML-test` and `ExpensiMark-HTMLToText-test`

1. What tests did you perform that validates your changed worked?
Added new test cases that auto link local urls.

# QA
1. What does QA need to do to validate your changes?
Run the new test cases

1. What areas to they need to test for regressions?
Check if urls are getting auto linked correctly

### CODE REVIEW

**Comment by mountiny** [comment]
@Santhosh-Sellavel any chance you can test this in App?

**Review by mountiny — APPROVED** [review]
Thanks @allroundexperts looking good, is there an easy way to test this with the App before merging?

File: lib/Url.js [hunk 1]
```diff
@@ -11,6 +11,9 @@ const URL_REGEX = `(${URL_WEBSITE_REGEX}${URL_PATH_REGEX}(?:${URL_PARAM_REGEX}|$
 
 const URL_REGEX_WITH_REQUIRED_PROTOCOL = URL_REGEX.replace(`${URL_PROTOCOL_REGEX}?`, URL_PROTOCOL_REGEX);
 
+const LOOSE_URL_WEBSITE_REGEX = `${URL_PROTOCOL_REGEX}([-\\w]+(\\.[-\\w]+)*)(?:\\:${ALLOWED_PORTS}|\\b|(?=_))`;
+const LOOSE_URL_REGEX = `(${LOOSE_URL_WEBSITE_REGEX}${URL_PATH_REGEX}(?:${URL_PARAM_REGEX}|${URL_FRAGMENT_REGEX})*)`;
+
```
Comment (Santhosh-Sellavel) [inline lib/Url.js:16]: Can we include the LOOSE_URL_REGEX along with the exported `URL_REGEX` as OR from here?

@allroundexperts

File: lib/Url.js [hunk 2]
```diff
@@ -11,6 +11,12 @@ const URL_REGEX = `(${URL_WEBSITE_REGEX}${URL_PATH_REGEX}(?:${URL_PARAM_REGEX}|$
 
 const URL_REGEX_WITH_REQUIRED_PROTOCOL = URL_REGEX.replace(`${URL_PROTOCOL_REGEX}?`, URL_PROTOCOL_REGEX);
 
+const LOOSE_URL_WEBSITE_REGEX = `${URL_PROTOCOL_REGEX}([-\\w]+(\\.[-\\w]+)*)(?:\\:${ALLOWED_PORTS}|\\b|(?=_))`;
```
Comment (neil-marcellini) [inline lib/Url.js:14]: Can you explain this chunk of the regex please? How did you come up with it and what is it supposed to do?
```javascript
([-\\w]+(\\.[-\\w]+)*)
```

My understanding is that it should match the domain name, however it isn't explained that way on [regex101](https://regex101.com/r/1EKXMm/1). Maybe I'm missing something...

**Review by neil-marcellini — COMMENTED** [review]
The tests look good but I'm a bit confused about the regex. I think you should also include test steps in App.

**Review by mountiny — APPROVED** [review]
@allroundexperts could you show tests with the App please?

File: lib/Url.js (hunk 2 above)
Comment (neil-marcellini) [reply lib/Url.js:14]: Ah got it thanks!

> You need to replace the `\\` with `\` when testing it on `regex101`

Can you explain why?

**Review by neil-marcellini — APPROVED** [review]

File: lib/ExpensiMark.js [hunk 3]
```diff
@@ -454,7 +454,7 @@ export default class ExpensiMark {
             if (abort) {
                 replacedText = replacedText.concat(textToCheck.substr(match.index, (match[0].length)));
             } else {
-                const urlRegex = new RegExp(`^${URL_REGEX}$`, 'i');
+                const urlRegex = new RegExp(`^${LOOSE_URL_REGEX}$|^${URL_REGEX}$`, 'i');
```
Comment (Santhosh-Sellavel) [inline lib/ExpensiMark.js:457]: Why can't we use MARKDOWN_URL_REGEX here?

**Review by neil-marcellini — APPROVED** [review]

**Review by Santhosh-Sellavel — APPROVED** [review]
Looks good to me, tests well. Will continue on App PR thanks!

cc: @neil-marcellini @mountiny

**Comment by Santhosh-Sellavel** [comment]
bump @mountiny or @neil-marcellini

**Review by mountiny — APPROVED** [review]
Thanks everyone!

File: lib/Url.js [hunk 4]
```diff
@@ -11,6 +11,12 @@ const URL_REGEX = `((${URL_WEBSITE_REGEX})${URL_PATH_REGEX}(?:${URL_PARAM_REGEX}
...
+const LOOSE_URL_REGEX = `((${LOOSE_URL_WEBSITE_REGEX})${URL_PATH_REGEX}(?:${URL_PARAM_REGEX}|${URL_FRAGMENT_REGEX})*)`;
```
Comment (allroundexperts) [inline lib/Url.js:15]: We allowed the URL to start by `hyphen` which is not allowed according to the [RFC1034](https://datatracker.ietf.org/doc/html/rfc1034). This caused https://github.com/Expensify/App/issues/27080

**Pull request merged**